### PR TITLE
Add `--adopt` switch

### PR DESCRIPTION
**Pull request by ThatsJustCheesy**
Allows `brew install` & co. to adopt existing cask artifacts that are identical to those being installed.

Discussion: #14006

- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

New tests are in `test/cask/artifact/app_spec.rb`. I can write more if requested.

Example of usage:

```
$ brew install textmate
==> Downloading https://github.com/textmate/textmate/releases/download/v2.0.23/TextMate_2.0.23.tbz
# …
==> Installing Cask textmate
==> Purging files for version 2.0.23 of Cask textmate
Error: It seems there is already an App at '/Applications/TextMate.app'.

$ brew install --adopt textmate
==> Downloading https://github.com/textmate/textmate/releases/download/v2.0.23/TextMate_2.0.23.tbz
# …
==> Installing Cask textmate
==> Adopting existing App at '/Applications/TextMate.app'
==> Linking Binary 'mate' to '/opt/homebrew/bin/mate'
🍺  textmate was successfully installed!

$ brew info textmate
==> textmate: 2.0.23 (auto_updates)
https://macromates.com/
/opt/homebrew/Caskroom/textmate/2.0.23 (6.2KB)
From: https://github.com/Homebrew/homebrew-cask/blob/HEAD/Casks/textmate.rb
# …
```

When the app at the target path differs:

```
$ brew install --adopt textmate
==> Downloading https://github.com/textmate/textmate/releases/download/v2.0.23/TextMate_2.0.23.tbz
# …
==> Installing Cask textmate
==> Adopting existing App at '/Applications/TextMate.app'
==> Purging files for version 2.0.23 of Cask textmate
Error: It seems the existing App is different from the one being installed.

$ brew install --adopt --verbose textmate
==> Downloading https://github.com/textmate/textmate/releases/download/v2.0.23/TextMate_2.0.23.tbz
# …
==> Verifying checksum for cask 'textmate'
==> Installing Cask textmate
# …
==> Adopting existing App at '/Applications/TextMate.app'
/usr/bin/diff -rq /opt/homebrew/Caskroom/textmate/2.0.23/TextMate.app /Applications/TextMate.app
Only in /Applications/TextMate.app: Icon
==> Purging files for version 2.0.23 of Cask textmate
Error: It seems the existing App is different from the one being installed.

$
```

Supplying both `--adopt` and `--force`:

```
$ brew install --adopt --force textmate
Usage: brew install [options] formula|cask [...]

# …
      --adopt                      Adopt existing artifacts in the destination
                                   that are identical to those being installed.
                                   Cannot be combined with --force.

# …

Error: Invalid usage: Options --adopt and --force are mutually exclusive.
```

### CODE REVIEW

**Comment by apainintheneck** [comment]
I really like this idea but I'm confused at what the intended use cases are for `brew reinstall --cask --adopt` and `brew upgrade --cask --adopt`.

**Comment by ThatsJustCheesy** [comment]
@apainintheneck Thanks for the question. I initially wasn't sure whether to add it to `reinstall` and `upgrade`. They do go through the same code paths as `install`, which is why I considered it at all. But of course, "adopting" as a concept only really makes sense when **new** casks are installed.

I came up with the following use cases to justify the changes to myself:

- `brew reinstall` works the same way as `brew install` for casks that aren't currently installed. e.g.,

   ```
   $ brew reinstall textmate
   ==> Downloading https://github.com/textmate/textmate/releases/download/v2.0.23/TextMate_2.0.23.tbz
   # …
   ==> Installing Cask textmate
   ==> Purging files for version 2.0.23 of Cask textmate
   Error: It seems there is already an App at '/Applications/TextMate.app'.

   $
   ```

   so `--adopt` meaningfully affects the behaviour in this case.
- I think(?) cask dependency changes can technically make `brew upgrade` install new casks. If that's true, then `--adopt` would apply. Admittedly, this is a very niche edge case.

**Comment by apainintheneck** [comment]
Thanks for explaining that to me.

With `brew reinstall` I can see the reasoning behind adding the `--adopt` option but I guess I wonder if it would be confusing to users. Normally, like you already mentioned, this command either uninstalls the current version and installs the most recent one or, if the package isn't already installed, it installs it. In this case, the adopted cask never gets uninstalled and the most recent version doesn't get installed either. I guess I would expect it to link the `.app` to a cask and then install the current version in it's place but I'm not sure how useful that would be.

I'm not 100% sure (hopefully someone else can chime in on this) but I don't think `brew upgrade` works like that.

**Comment by MikeMcQuaid** [comment]
> I initially wasn't sure whether to add it to `reinstall` and `upgrade`. They do go through the same code paths as `install`, which is why I considered it at all. But of course, "adopting" as a concept only really makes sense when **new** casks are installed.

Yeh, please remove `upgrade` and `reinstall`. We can add `reinstall` later if anyone requests it. We should not add `upgrade`.

File: Library/Homebrew/cask/artifact/moved.rb [hunk 1]
```diff
@@ -34,8 +34,33 @@ def summarize_installed
...
+        unless source.exist?
+          raise CaskError, "It seems the #{self.class.english_name} source '#{source}' is not there."
+        end
+
         if Utils.path_occupied?(target)
+          if adopt
+            ohai "Adopting existing #{self.class.english_name} at '#{target}'"
+            same = command.run(
+              "/usr/bin/diff",
+              args:         ["-rq", source, target],
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/artifact/moved.rb:47]:
```suggestion
              args:         ["--recursive", "--brief", source, target],
```

File: Library/Homebrew/cask/artifact/moved.rb [hunk 2]
```diff
@@ -34,8 +34,33 @@ def summarize_installed
...
+              print_stdout: verbose,
+            ).success?
+
+            unless same
+              raise CaskError,
+                    "It seems the existing #{self.class.english_name} is different from " \
+                    "the one being installed."
+            end
+
+            # Simulate moving the source to the target location
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/artifact/moved.rb:58]:
```suggestion
            # Remove the source as we don't need to move it to the target location
```

**Review by MikeMcQuaid — COMMENTED** [review]
Looks good so far! I'm wonder if something like "reuse" may make more sense as a name. @Homebrew/brew @Homebrew/cask any thoughts?

**Comment by SMillerDev** [comment]
I think adopting an existing package is the better term. Reuse sounds like you'd have to specify a tool to (re)use

**Comment by ThatsJustCheesy** [comment]
> Yeh, please remove `upgrade` and `reinstall`. We can add `reinstall` later if anyone requests it. We should not add `upgrade`.

Done!

**Review by MikeMcQuaid — APPROVED** [review]
Thanks so much for your first contribution! Without people like you submitting PRs we couldn't run this project. You rock, @ThatsJustCheesy!

**Pull request merged**